### ISSUE TITLE
Advertise SabreDAV PATCH support

### ISSUE THREAD

**Issue by simaoafonso-pwt** [nextcloud/server#15923]
Advertise SabreDAV PATCH support, so that the files can be mounted using FUSE on Linux, with https://github.com/miquels/webdavfs.

See https://github.com/miquels/webdavfs/issues/8#issuecomment-474357271

> The [SabreDav PATCH specification](http://sabre.io/dav/http-patch/) says that sabredav-partialupdate needs to be advertised in the DAV: header.

**Comment by kesselb** [comment]
cc @icewind1991 @rullzer

**Comment by jyrkij** [comment]
Any progress on this? Does NC use the SabreDav plugin in question?

**Comment by kesselb** [comment]
> Does NC use the SabreDav plugin in question?

The plugin / interface for using SabreDav PATCH is not implemented.

**Comment by simaoafonso-pwt** [comment]
Bump

**Comment by kesselb** [comment]
Feel free to submit a patch :) I'm not sure if announcing it is enough. Server must be able to handle chunked uploads in that format.

**Comment by simaoafonso-pwt** [comment]
> Feel free to submit a patch :) I'm not sure if announcing it is enough. Server must be able to handle chunked uploads in that format.

You are correct of course. Since SabreDAV itself supports this, I assumed this was simply toggling a switch. I don't enough PHP to take on this.

- Docs: https://sabre.io/dav/http-patch/
- Sample Code: https://github.com/nextcloud/3rdparty/blob/master/sabre/dav/lib/DAV/FSExt/File.php

**Comment by PVince81** [comment]
Some external storages might not support "fseek" so would require a fallback to make it first download the file, then patch it, then reupload, that could become messy. Or it would need to return 501 then just for those storages.

For a local storage I guess it could work fine. The implementation would be more complex as there are more things to consider, like firing hooks, etc. See how the regular PUT is implemented in NC: https://github.com/nextcloud/server/blob/master/apps/dav/lib/Connector/Sabre/File.php#L125
There's a lot going on there.

**Comment by szaimen** [comment]
cc @nextcloud/server do we want this?

**Comment by blizzz** [comment]
looks nice, but comes with the rat tail as pointed out by @PVince81

**Comment by cbertelli** [comment]
It's dangerous, but meanwhile you could add a mounting option to enable sabreDAV behaviour even if not advertised.

**Comment by muety** [comment]
Any ideas when this fix might get implemented?

**Comment by aguaviva** [comment]
> Bump.
> 
> https://github.com/nextcloud/server/blob/b219eadb8f3f459f2c71e937dfd3542f0792da8e/apps/dav/lib/DAV/Sharing/Plugin.php#L76
> 
> This sounds like a simple fix: Adding `sabredav-partialupdate` to that array.

Any updates on this? Is this really the fix for local storage? I'll be happy to test it.

**Comment by PVince81** [comment]
As far as I see in the README https://github.com/miquels/webdavfs, FUSE itself will send partial updates over the network, 
so it's not as easy as advertising support without actually implementing it

To implement partial update see https://github.com/nextcloud/server/issues/15923#issuecomment-757936355
It would either require adding this inefficient workaround in the DAV layer of Nextcloud, and write performance will likely suffer

Or implement it in a clean way, but this requires extending the storage API on various layers in Nextcloud to support writing a chunk in the middle of a file. Then, every storage implementation (Local, S3, SMB, etc) could either implement it directly if supported by their APIs, or use the inefficient workaround (download, then patch, then reupload).

Also, need to make sure that server side encryption and various storage wrappers (ex: access control, group folders) are also working correctly with this addition.

In any case, this sounds like it would be a bigger project involving a lot of code additions in many layers.

**Comment by simaoafonso-pwt** [comment]
Is this a per-server setting? Is it possible to advertise support for the extension on the local storage folder? At least that would take care of the simplest use case.

**Comment by PVince81** [comment]
> Is this a per-server setting? Is it possible to advertise support for the extension on the local storage folder? At least that would take care of the simplest use case.

as far as I understand it's a per-route setting on the DAV layer
you'd need to add a check to see what storage type is used on that path

but this is not enough, a new method `putRange()` needs to be implemented in https://github.com/nextcloud/server/blob/master/apps/dav/lib/Connector/Sabre/File.php#L123 otherwise write operations with webdavfs/fuse will likely fail as PATCH is not implemented. Ref https://sabre.io/dav/http-patch/

**Comment by solracsf** [comment]
As this is not implemented, why is Desktop client trying to do such calls?

```
PATCH "/remote.php/dav/files/user/18_Energie/2021_pacte-15/7_Finance/Copie%20de%20Export_2021-01_14.10.22.xls" code=501 ua="Mozilla/5.0 (Windows) mirall/3.6.1stable-Win64 (build 20221018) (Nextcloud, windows-10.0.19044 ClientArchitecture: x86_64 OsArchitecture: x86_64)"
```

**Comment by kesselb** [comment]
https://sabre.io/dav/http-patch/ means to replace a given part of a file.

Your request is most likely a request to update the checksum of a file: https://github.com/nextcloud/server/blob/66c33fd9b8e21dee5332ea8df8b9b4b718709711/apps/dav/lib/Connector/Sabre/ChecksumUpdatePlugin.php#L36

**Comment by stefan123t** [comment]
Thanks @kesselb for pointing me to this issue :+1: 

> Some external storages might not support "fseek" so would require a fallback to make it first download the file, then patch it, then reupload, that could become messy. Or it would need to return 501 then just for those storages.
> 
> For a local storage I guess it could work fine. The implementation would be more complex as there are more things to consider, like firing hooks, etc. See how the regular PUT is implemented in NC: https://github.com/nextcloud/server/blob/master/apps/dav/lib/Connector/Sabre/File.php#L125 There's a lot going on there.

So these are the main things to be considered for SabreDAV PATCH support in Nextcloud:

* if local / remote storage supports fseek:
  - advertise `DAV: sabredav-partialupdate` in OPTIONS / PROPFIND response.

* implement PATCH method in https://github.com/nextcloud/server/blob/master/apps/dav/lib/Connector/Sabre/File.php
  depending on 
  - `X-Update-Range: append` using `fopen($this->path, 'a')` to the end of the file and append PATCH body as requested for logfiles in #40968 as in `case 1:` or per **`default:`**
   - `X-Update-Range: bytes=3-6` using `fopen($this->path, 'c')` and fseek to change from byte 3 and insert into / replace the file content until byte 6 using the PATCH body as in `case 2:` and `case 3:`

This has been implemented in https://github.com/sabre-io/dav/blob/a74464d583af69d72106ad48820b52bb44144e43/lib/DAV/FSExt/File.php#L64-L91

```php
    public function patch($data, $rangeType, $offset = null)
    {
        switch ($rangeType) {
            case 1:
                $f = fopen($this->path, 'a');
                break;
            case 2:
                $f = fopen($this->path, 'c');
                fseek($f, $offset);
                break;
            case 3:
                $f = fopen($this->path, 'c');
                fseek($f, $offset, SEEK_END);
                break;
            default:
                $f = fopen($this->path, 'a');
                break;
        }
        if (is_string($data)) {
            fwrite($f, $data);
        } else {
            stream_copy_to_stream($data, $f);
        }
        fclose($f);
        clearstatcache(true, $this->path);


        return $this->getETag();
    }
```

In case the local / remote backend does not support fseek, we should return `501 Not Implemented` as @PVince81 suggested.

I am not sure if I understood his comment on putRange correct here, but I think he would prefer to factor-out the second `X-Update-Range: bytes=N-M` into a `putRange()` method which is only called in case the fseek command is understood by the local / remote filesystem backend for `$this->path` ?

The SabreDAV PartialUpdate Plugin is already implemented in https://github.com/sabre-io/dav/blob/a74464d583af69d72106ad48820b52bb44144e43/lib/DAV/PartialUpdate/Plugin.php so what kind of changes for the integration or glue code would still be needed to facilitate this with Nextcloud ?

**Comment by stefan123t** [comment]
Congratulations @PVince81 & @kesselb for getting the Nextcloud 30 release out of the door!

Can we get a short reply on the status / plans for this issue please ?

As mentioned I would like to use the simplest option of the easy use case only,
i.e. storage on a **local filesystem** (which **supports fseek**) and **only append** to the file.
For me a small IoT ESP32 based device (Solar PV inverter OpenSource DTU) should simply be appending its production data to an CSV file. 
I do not want to implement some G**gle Sheets API to store OpenSource data into the data scrapers behemoth, I would like to store it on my Nextcloud instance.

1. So as @simaoafonso-pwt has suggested we should advertise the `sabredav-partialupdate` in case the current path is on local storage, i.e. supports fseek.

2. This conditional advertising might be a bit tricky as @PVince81 has hinted. 
We would somehow need to query the current path's file storage backend implementation (and its potential fseek support). 
Is there an example or should we simply advertise it always for starters ?

3. Then only PATCH method has to be implemented / extended to allow appending the new content to the end of the file, as in the above SabreDAV `File.php`. 
I am a bit lost at the moment: How and where would that be best approached in the case of Nextcloud Code structure and PHP classes ? 

4. The necessary PUT, GET, PROPFIND, **OPTIONS** and **PATCH** method / commands we would likely need to call I have documented in #40968 if we need some test cases for such an implementation.

Bonus question: would the PATCH range types 2 and 3 which are using / "changing" ` fopen($this->path, 'c');` also apply to / ease the implementation of Delta-sync as described in #417 ?
